### ISSUE TITLE
pkcs11 support for SSH Server

### ISSUE THREAD

**Issue by bmoe3rd** [ronf/asyncssh#358]
Hello dear asyncssh developers,

this is a follow up to a question i asked on asyncssh-users:
https://groups.google.com/g/asyncssh-users/c/YzmgmF1geMU

i would like to contribute pkcs11 token support for the **server** and discuss how to go about it.
Also i would like to provide the certificate for the private key as a **SSH** Certificate from disk.

I skimmed over the sources and believe, that the easiest way to go about this would be to:
1. Modify SSHServerConnectionOptions so that server_host_keys may be a list of keypairs
2. Implement SSHPKCS11KeyPair.set_certificate to accept a SSH Certificate

I personally am totally sufficient with one private key, so matching keys with certificates is easy for me.
I don't know how you would go about it for multiple keys.

What do you think?

Best regards,
Birger Möllering

**Comment by ronf** [comment]
Thanks for your interest in AsyncSSH and in expanding its PKCS#11 capabilities!

The first item in your list here should already be the case, I think. I did a quick test where I used load_pkcs11_keys() to load a set of SSHPKCS11KeyPair objects and then passed in objects from that list in the server_host_keys argument of a call to asyncssh.listen() and I was able to get it to use that key pair for its host key. It can already handle certificates as well, but only X.509 certificates loaded onto the PIV token (and only when the client supports using X.509 certificates for SSH, which isn't available in OpenSSH or many other SSH implementations).

The primary change would be the second item in your list, but it's pretty straightforward to do that. In fact, I wrote a prototype of that tonight which I'll try out soon.

As for matching up the keys with the certificates, this can be done by comparing the public_data bytes of the key and the certificate's key. In fact, there's already code in load_keypairs() which can do this matching for you, in cases where you pass in separate lists of keys and certificates rather than passing in tuples of (key, cert). I think you might be able to leverage that here if you pass in a list of PKCS#11 keypairs in server_host_keys and also pass in server_host_certs. It looks like it might need a small tweak to work on keypairs that already have a certificate associated with them, though. We need to compare the public_data of the original key, not the public_data associated with the existing cert. With that change plus an implementation of set_certificate() in SSHPKCS11KeyPair, I think you could be good to go.

The other option for passing in a certificate would be to pass in a tuple of (pkcs11_keypair, cert) as elements in the server_host_keys list. You'd have to match up the certs yourself in that case, and you'd still need an implementation of set_certificate() before that will work, but once set_certificate() is in place either form of specifying the certificates should be available to you.

I'll get back to you after I try out the new set_certificate() code.

**Comment by ronf** [comment]
Ok - I have something working, but it needs polishing. I think I can rework the SSHPKCS11KeyPair class to eliminate some of the code duplication, and I also need to add some new unit tests. However, if you'd like to experiment with it in the meantime, here's a diff:

```patch
diff --git a/asyncssh/pkcs11.py b/asyncssh/pkcs11.py
index 3140e78..23ec3e4 100644
--- a/asyncssh/pkcs11.py
+++ b/asyncssh/pkcs11.py
@@ -65,6 +65,7 @@ if pkcs11_available:

             self._session = session
             self._key = key
+            self._pubkey = cert.key if cert else pubkey
             self._cert = cert

             if cert:
@@ -79,8 +80,29 @@ if pkcs11_available:
         def __del__(self):
             self._session.close()

+        @property
+        def key_public_data(self):
+            """Return public data of key in this keypair"""
+
+            return self._pubkey.public_data
+
         def set_certificate(self, cert):
-            """Set certificate not applicable to PKCS#11 keys"""
+            """Set alternate certificate to use with this key"""
+
+            if cert.key.public_data != self._pubkey.public_data:
+                raise ValueError('Certificate key mismatch')
+
+            self._cert = cert
+            self.algorithm = cert.algorithm
+            self.public_data = cert.public_data
+
+            if cert.is_x509_chain:
+                self.sig_algorithm = cert.algorithm
+            else:
+                self.sig_algorithm = self._pubkey.algorithm
+
+            self.sig_algorithms = cert.sig_algorithms
+            self.host_key_algorithms = cert.host_key_algorithms

         def set_sig_algorithm(self, sig_algorithm):
             """Set the signature algorithm to use when signing data"""
diff --git a/asyncssh/public_key.py b/asyncssh/public_key.py
index c0dc1dc..1b4643b 100644
--- a/asyncssh/public_key.py
+++ b/asyncssh/public_key.py
@@ -1831,6 +1831,12 @@ class SSHKeyPair:
         self.set_comment(comment)
         self._filename = filename

+    @property
+    def key_public_data(self):
+        """Return public data of key in this keypair"""
+
+        return self._key.public_data
+
     def get_key_type(self):
         """Return what type of key pair this is

@@ -3202,7 +3208,12 @@ def load_keypairs(keylist, passphrase=None, certlist=(), skip_public=False):
             raise saved_exc # pylint: disable=raising-bad-type

         if not certs:
-            cert = certdict.get(key.public_data)
+            if isinstance(key, SSHKeyPair):
+                pubdata = key.key_public_data
+            else:
+                pubdata = key.public_data
+
+            cert = certdict.get(pubdata)

             if cert and cert.is_x509:
                 cert = SSHX509CertificateChain.construct_from_certs(certlist)
```

To use this, you'd do something like:

```python
    host_key = asyncssh.load_pkcs11_keys(provider, pin)[0]
    await asyncssh.create_server(MySSHServer, '', 8022,
                                 server_host_keys=[host_key],
                                 server_host_certs=[host_cert_filename],
                                 ...other_create_server_arguments...)
```

Alternately, you could do something like:
```python
    host_key = asyncssh.load_pkcs11_keys(provider, pin)[0]
    await asyncssh.create_server(MySSHServer, '', 8022,
                                 server_host_keys=[(host_key, host_cert_filename)],
                                 ...other_create_server_arguments...)
```

**Comment by bmoe3rd** [comment]
Awesome!
I applied the diffs to the 2.5.0 branch and wrote a little Test Server for it.
The connection was closed silently by the server, whenever i tried to connect to it with openssh.
The cause is a `pkcs11.exceptions.MechanismInvalid` Exception.
Most probably the key in the tpm does not match with any mechanisms listed in pkcs11.py !?
Currently i am using a key with following template:
```
name-alg:
  value: sha256
  raw: 0xb
attributes:
  value: fixedtpm|fixedparent|sensitivedataorigin|userwithauth|restricted|sign
  raw: 0x50072
type:
  value: rsa
  raw: 0x1
exponent: 65537
bits: 2048
scheme:
  value: rsassa
  raw: 0x14
scheme-halg:
  value: sha256
  raw: 0xb
sym-alg:
  value: null
  raw: 0x10
sym-mode:
  value: (null)
  raw: 0x0
sym-keybits: 0
```

Most probably it is time for me to read about signing schemes and tpm 2.0 again.
Must the private key be able to decrypt or is signing enough?

The Server i used:
```python
#!/usr/bin/env python3

import asyncio, asyncssh, sys

asyncssh.set_log_level(3)
asyncssh.set_debug_level(3)

keys = asyncssh.load_pkcs11_keys(
    "/usr/lib/x86_64-linux-gnu/pkcs11/libtpm2_pkcs11.so.0.0.0",
    pin="userpin",
    load_certs=False,
    token_label="sshtoken",
    key_label="signingkey"
)

def handle_client(process):
    username = process.get_extra_info('username')
    process.stdout.write(f"Welcome to my SSH server, {username}!\n")
    process.exit(0)

class MySSHServer(asyncssh.SSHServer):
    def connection_made(self, conn):
        print(f'made connection with {conn}')
        self._conn = conn

    def connection_lost(self, exc):
        if exc:
            print(f'SSH connection error: ' + str(type(exc)), file=sys.stderr)
        else: print('SSH connection closed')

    def begin_auth(self, username):
        print(f"begin auth with {username}")
        return True


async def start_server():
    await asyncssh.create_server(
        MySSHServer,
        "",
        8022,
        server_host_keys=[
            (keys[0], "./ssh.bak/drone-cert.pub")
        ],
        authorized_client_keys="./ssh.bak/ca.pub",
    )

loop = asyncio.get_event_loop()
try:
    loop.run_until_complete(start_server())
except (OSError, asyncssh.Error) as exc:
    sys.exit(f"Error stating server: {exc}")
loop.run_forever()
```

**Comment by ronf** [comment]
Your server code looks fine. I agree it's probably related to the key type you are choosing. Specifically, AsyncSSH currently supports the following key types and mechanisms:

```python
    encoders = {KeyType.RSA: encode_rsa_public_key,
                KeyType.EC:  encode_ec_public_key}

    mechanisms = {b'ssh-rsa':                Mechanism.SHA1_RSA_PKCS,
                  b'rsa-sha2-256':           Mechanism.SHA256_RSA_PKCS,
                  b'rsa-sha2-512':           Mechanism.SHA512_RSA_PKCS,
                  b'ssh-rsa-sha224@ssh.com': Mechanism.SHA224_RSA_PKCS,
                  b'ssh-rsa-sha256@ssh.com': Mechanism.SHA256_RSA_PKCS,
                  b'ssh-rsa-sha384@ssh.com': Mechanism.SHA384_RSA_PKCS,
                  b'ssh-rsa-sha512@ssh.com': Mechanism.SHA512_RSA_PKCS,
                  b'rsa1024-sha1':           Mechanism.SHA1_RSA_PKCS,
                  b'rsa2048-sha256':         Mechanism.SHA256_RSA_PKCS,
                  b'ecdsa-sha2-nistp256':    Mechanism.ECDSA_SHA256,
                  b'ecdsa-sha2-nistp384':    Mechanism.ECDSA_SHA384,
                  b'ecdsa-sha2-nistp521':    Mechanism.ECDSA_SHA512}
```

Just signing should be good enough here, at least for most of the RSA key exchange options. I think encrypt/decrypt is only needed for 'rsa1024-sha1' and 'rsa2048-sha256', and while they are listed in the mechanism table here I'm not sure I have support yet for using the KexRSA key exchange with PKCS11 keys.

I don't see a padding algorithm configured in your TPM config, so I'm wondering if maybe it is trying to use something other than PKCS for that.

What do you see when you print out keys[0]._key, keys[0]._pubkey, and keys[0].sig_algorithm?

I'm not familiar with libtpm2, but I see you have it set for a hashing alg of SHA256. That's supported by AsyncSSH, but the default sig_algorithm is 'ssh-rsa' (SHA-1), not 'rsa-sha2-256' or 'rsa-sha2-512', so there may be an issue depending on whether the client supports those or not.

Have you been able to get this to work with AsyncSSH as a client key, before these changes? The only thing I have to test against here right now is a Yubikey.

**Comment by bmoe3rd** [comment]
> I don't see a padding algorithm configured in your TPM config, so I'm wondering if maybe it is trying to use something other than PKCS for that.

Oh well, the explanations about what a padding algorithm is on the web are easy enough,
but nevertheless i am out of my depth here. xD

> What do you see when you print out keys[0]._key, keys[0]._pubkey, and keys[0].sig_algorithm?
```
keys[0]._key: <PrivateKey label='signingkey' id='30783831303130303032' 2048-bit RSA>
keys[0]._pubkey: <asyncssh.rsa._RSAKey object at 0x7fd656308ca0>
keys[0].sig_algorithm: b'ssh-rsa'
```
shouldn't it be `rsa-sha2-256` here?

> Have you been able to get this to work with AsyncSSH as a client key, before these changes?

The private Key that i use, is generated **on** the tpm,
i merely get the public portion from it -> sign it with the ca private key -> and get it back to the other machine
**But** it worked with the ssh-agent from openssh as a HostKeyAgent, merely the pairing with the certificate failed!
I don't think it is possible to create private keys somwhere else and store them on/with the tpm later (though i may be wrong).
The tpm2-pkcs11 library is _only_ another interface to the tpm.

> The only thing I have to test against here right now is a Yubikey.

If you are working on a moderately recent x86_64 pc, you most probably have a tpm 2.0 device already!
It's not an easy matter to start with though! That chip is a piece of work and i seriously suggest not to start with it if you don't have time to burn. Your entry point to it might be a bit better than mine though ;)
But if you are Interessted i can give you a hand.

**Comment by ronf** [comment]
> Oh well, the explanations about what a padding algorithm is on the web are easy enough,
> but nevertheless i am out of my depth here. xD

I know how to deal with this at the PyCA (Python cryptography library) and PKCS#11 Mechanism level, but not specifically how to handle this in libtpm2.

> What do you see when you print out keys[0]._key, keys[0]._pubkey, and keys[0].sig_algorithm?
> 
> > keys[0]._key: <PrivateKey label='signingkey' id='30783831303130303032' 2048-bit RSA>
> > keys[0]._pubkey: <asyncssh.rsa._RSAKey object at 0x7fd656308ca0>
> > keys[0].sig_algorithm: b'ssh-rsa'
> 
> shouldn't it be rsa-sha2-256 here?

If you print out keys[0].sig_algorithms, you should see a list of all the signature algorithms which are supported by that key. There's a separate step when opening the connection where the two sides negotiate signature algorithms, and "rsa-sha2-256 " and "rsa-sha2-512" are the most preferred. So, if the client supports either of those, they should be chosen over the older "ssh-rsa". There's a call to set_sig_algorithm() on the key which should do this. You could check to see if that's happening in your case before sign() is called.

If you have the ability to set up an ECDSA key, you might want to try that. It doesn't have the extra complication of there being multiple signature algorithms for a single key type.

You could also print out "mechanisms[sig_algorithm]" in the sign() function to see what mechanism it is requesting from the PKCS#11 provider.

You can also run "pkcs11-tool --module _provider_ -M" to see what mechanisms are supported.. On libykcs11 from Yubico, I get:

```
Using slot 0 with a present token (0x0)
Supported mechanisms:
  RSA-PKCS-KEY-PAIR-GEN, keySize={1024,2048}, hw, generate_key_pair
  RSA-PKCS, keySize={1024,2048}, hw, encrypt, decrypt, sign, verify
  RSA-PKCS-PSS, keySize={1024,2048}, hw, sign, verify
  RSA-PKCS-OAEP, keySize={1024,2048}, hw, encrypt, decrypt
  RSA-X-509, keySize={1024,2048}, hw, encrypt, decrypt, sign, verify
  SHA1-RSA-PKCS, keySize={1024,2048}, hw, sign, verify
  SHA256-RSA-PKCS, keySize={1024,2048}, hw, sign, verify
  SHA384-RSA-PKCS, keySize={1024,2048}, hw, sign, verify
  SHA512-RSA-PKCS, keySize={1024,2048}, hw, sign, verify
  SHA1-RSA-PKCS-PSS, keySize={1024,2048}, hw, sign, verify
  SHA256-RSA-PKCS-PSS, keySize={1024,2048}, hw, sign, verify
  SHA384-RSA-PKCS-PSS, keySize={1024,2048}, hw, sign, verify
  SHA512-RSA-PKCS-PSS, keySize={1024,2048}, hw, sign, verify
  ECDSA-KEY-PAIR-GEN, keySize={256,384}, hw, generate_key_pair
  ECDSA, keySize={256,384}, hw, sign, verify
  ECDSA-SHA1, keySize={256,384}, hw, sign, verify
  ECDSA-SHA224, keySize={256,384}, hw, sign, verify
  ECDSA-SHA256, keySize={256,384}, hw, sign, verify
  ECDSA-SHA384, keySize={256,384}, hw, sign, verify
  ECDH1-DERIVE, keySize={256,384}, hw, derive
  SHA-1, digest
  SHA256, digest
  SHA384, digest
  SHA512, digest
```

> The private Key that i use, is generated on the tpm, I merely get the public portion from it -> sign it with the ca private key -> and get it back to the other machine

Yeah - this is standard practice for devices like this. Even with the Yubikey, it only provides a way to extract the public key corresponding to the private key it has internal to it. You can tell it to generate a new private key at any time, but you can only retrieve the public key. You can create a certificate on the host and store it on the token, but it has to be an X.509 certificate.

> But it worked with the ssh-agent from openssh as a HostKeyAgent, merely the pairing with the certificate failed!

Huh - I didn't realize that you can tell OpenSSH to use ssh-agent to hold onto your host keys. How does that work? On AsyncSSH, I currently only contact the ssh-agent for client keys.

For client keys, I know ssh-agent supports certificates as well, but you have to provide the cert at the time you load a key into it.

> If you are working on a moderately recent x86_64 pc, you most probably have a tpm 2.0 device already!
> It's not an easy matter to start with though! That chip is a piece of work and i seriously suggest not to start with it if you don't have time to burn. Your entry point to it might be a bit better than mine though ;)
> But if you are Interessted i can give you a hand.

I appreciate the offer. I have pretty recent hardware here (2019/2020), but all of the systems are Macs, and from what I can see libtpm2 doesn't seem to support the Mac. I have Windows and Linux running here, but only as virtual machines, so that wouldn't help for something like a TPM.

There is a version of "SoftHSM" for the Mac that I did tinker with at one point, but I didn't manage to get it to work at the time and never got around to trying it again since I had gotten the Yubikey working as both a FIDO and PIV token by then.

**Comment by ronf** [comment]
I finally got a chance to get back to this, and commit 2fb89a7 in the "develop" branch now contains a generalized version of this fix which not only allows alternate certificates to be specified for PKCS11 keys, but also allows this for SSH agent keys. This means you can use an OpenSSH certificate with an agent key without loading that cert into the agent, or even use an X.509 cert with an agent key (which the regular OpenSSH agent doesn't support).

I'm not sure if you were ever able to work out the issues with your HSM config, but if so I'm hoping these changes will provide what you need.

**Comment by ronf** [comment]
This support is now available in AsyncSSH 2.6.0.
